hydro_lang/src/stream.rs

```diff
@@ -57,6 +57,36 @@ impl MinOrder<TotalOrder> for NoOrder {
     type Min = NoOrder;
 }
 
+/// Marks the stream as having deterministic message cardinality, with no
+/// possibility of duplicates.
+pub struct ExactlyOnce;
+
+/// Marks the stream as having non-deterministic message cardinality, which
+/// means that duplicates may occur, but messages will not be dropped.
+pub struct AtLeastOnce;
```

Comment:
using an empty enum will prevent the type from being instantiated (also applies to all the other marker types)